core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -197,28 +194,20 @@ public static Options defaultOptions() {
   private final int mMaxCapacity;
   private final int mMinCapacity;
 
-  // Tracks the resources that are available ordered by lastAccessTime (the first one is
+  // Tracks the resources that are available ordered by lastAccessTime (the head is
   // the most recently used resource).
+  // These are the resources that acquire() will take.
+  // This is always a subset of the other data structure mResources.
   @GuardedBy("mLock")
-  private TreeSet<ResourceInternal<T>> mResourceAvailable =
-      new TreeSet<>(new Comparator<ResourceInternal<T>>() {
-        @Override
-        public int compare(ResourceInternal<T> c1, ResourceInternal<T> c2) {
-          if (c1 == c2) {
-            return 0;
-          }
-          if (c1.mLastAccessTimeMs == c2.mLastAccessTimeMs) {
-            return c1.mIdentity - c2.mIdentity;
-          }
-          return (int) (c2.mLastAccessTimeMs - c1.mLastAccessTimeMs);
-        }
-      });
+  private final Deque<ResourceInternal<T>> mAvailableResources;
 
   // Tracks all the resources that are not closed.
-  @GuardedBy("mLock")
-  private HashMap<T, ResourceInternal<T>> mResources = new HashMap<>(32);
+  // The size of the map is guarded by mLock. The mLock should be acquired if you add/delete
```

Comment:
Is there a reason why you are reusing the existing `mLock`? `mLock` seems to be a lock that protects concurrent access to `mAvailableResources`. Shouldn't there be a separate lock that protects this `mResources` data structure?